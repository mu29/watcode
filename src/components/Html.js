/* eslint-disable react/no-danger */
import React from 'react'
import { PropTypes } from 'prop-types'
import { Helmet } from 'react-helmet'

const Html = ({ styles, state, children }) => {
  const helmet = Helmet.renderStatic()
  const htmlAttrs = helmet.htmlAttributes.toComponent()
  const bodyAttrs = helmet.bodyAttributes.toComponent()

  return (
    <html lang="ko" { ...htmlAttrs }>
      <head>
        { helmet.title.toComponent() }
        { helmet.meta.toComponent() }
        { helmet.link.toComponent() }
        <meta charSet="utf8" />
        <link href="/style.css" rel="stylesheet" type="text/css" />
        <link href="//spoqa.github.io/spoqa-han-sans/css/SpoqaHanSans-kr.css" rel="stylesheet" type="text/css" />
        <script defer src="https://use.fontawesome.com/releases/v5.0.13/js/all.js" integrity="sha384-xymdQtn1n3lH2wcu0qhcdaOpQwyoarkgLVxC/wZ5q7h9gHtxICrpcaSUfygqZGOe" crossOrigin="anonymous" />
        { styles }
      </head>
      <body { ...bodyAttrs }>
        <main id="app">{ children }</main>
        <script dangerouslySetInnerHTML={{ __html: state }} />
        <script src="/client.js" />
      </body>
    </html>
  )
}

Html.propTypes = {
  styles: PropTypes.node.isRequired,
  state: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
}

export default Html
