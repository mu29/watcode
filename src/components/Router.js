import React from 'react'
import { PropTypes } from 'prop-types'
import { Switch, Route, Redirect } from 'react-router-dom'
import routes from 'routes'

export default class Router extends React.Component {
  static propTypes = {
    isLoggedIn: PropTypes.bool,
  }

  static defaultProps = {
    isLoggedIn: false,
  }

  renderComponent = (route, isLoggedIn) => (props) => {
    if (!isLoggedIn && route.private) {
      return <Redirect to={{ pathname: '/login', state: { from: props.location } }} />
    }

    if (isLoggedIn && ['/login', '/'].includes(route.path)) {
      return <Redirect to="/profile" />
    }

    return <route.component { ...props } />
  }

  renderRoute = route => (
    <Route
      key={ route.path }
      path={ route.path }
      exact={ route.exact }
      render={ this.renderComponent(route, this.props.isLoggedIn) }
    />
  )

  render() {
    return (
      <Switch>
        { routes.map(route => this.renderRoute(route)) }
      </Switch>
    )
  }
}
