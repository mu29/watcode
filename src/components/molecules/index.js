export { default as Field } from './Field'
export { default as IconButton } from './IconButton'
export { default as Tab } from './Tab'
