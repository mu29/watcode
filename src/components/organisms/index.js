export { default as ArtworkList } from './ArtworkList'
export { default as Header } from './Header'
export { default as LoginForm } from './LoginForm'
export { default as ReduxField } from './ReduxField'
export { default as RegisterForm } from './RegisterForm'
export { default as SearchBar } from './SearchBar'
