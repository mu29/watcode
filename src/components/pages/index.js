export { default as HomePage } from './HomePage'
export { default as RankingPage } from './RankingPage'
export { default as BookmarkPage } from './BookmarkPage'
