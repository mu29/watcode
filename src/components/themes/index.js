export { default as basicTheme } from './basic'
