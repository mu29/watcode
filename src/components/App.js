import React from 'react'
import PropTypes from 'prop-types'
import Helmet from 'react-helmet'
import { injectGlobal, ThemeProvider } from 'styled-components'
import { authorizeActions, readBookmarksActions } from 'store/actions'
import { auth } from 'services/firebase'
import { basicTheme } from './themes'
import Router from './Router'

/* eslint-disable no-unused-expressions */
injectGlobal`
  * {
    box-sizing: border-box;
  }
  body {
    margin: 0;
  }
`

export default class App extends React.Component {
  static propTypes = {
    store: PropTypes.shape({
      dispatch: PropTypes.func,
      getState: PropTypes.func,
    }).isRequired,
  }

  componentWillMount() {
    const { store } = this.props
    auth.onAuthStateChanged((user) => {
      store.dispatch(authorizeActions.request({ user }))
      store.dispatch(readBookmarksActions.request())
    })
  }

  render() {
    const state = this.props.store.getState()
    const isLoggedIn = state.auth.user.email !== undefined
    return (
      <React.Fragment>
        <Helmet titleTemplate="watcode">
          <title>왓코드 - 오늘 볼 만화 추천해줘!</title>
          <meta charSet="UTF-8" />
          <meta name="description" content="" />
          <meta name="viewport" content="width=device-width, initial-scale=1.0" />
          <meta property="og:site_name" content="watcode" />
        </Helmet>
        <ThemeProvider theme={ basicTheme }>
          <Router isLoggedIn={ isLoggedIn } />
        </ThemeProvider>
      </React.Fragment>
    )
  }
}
