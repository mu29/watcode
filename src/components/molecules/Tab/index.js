import React, { Component } from 'react'
import { PropTypes } from 'prop-types'
import styled from 'styled-components'
import { Button } from 'components'
import { palette, borderColor } from 'services/style'

const getWeight = ({ selected }) => (selected ? 600 : 400)

const getSelectedColor = ({ selected, theme }) =>
  palette(selected ? 'gray.90' : 'gray.90')({ theme })

const getSelectedBorder = ({ selected, theme }) =>
  (selected ? `0.1875rem solid ${palette('yellow.default')({ theme })}` : 'none')

const TabContainer = styled.div`
  width: 100%;
  height: 100%;
`

const TabHeader = styled.div`
  display: flex;
`

const TabDivider = styled.div`
  height: 0.125rem;
  margin-top: -0.125rem;
  z-index: -1;
  background-color: ${borderColor};
`

const TabItem = styled(Button)`
  box-sizing: content-box;
  height: 2.5rem;
  padding: 0rem 0.75rem;
  font-weight: ${getWeight};
  color: ${getSelectedColor};
  background-color: transparent;
  border-radius: 0;
  border-bottom: ${getSelectedBorder};

  &:hover {
    color: ${getSelectedColor};
  }
`

class Tab extends Component {
  state = {
    index: 0,
  }

  handleTabChange = index => this.setState({ index })

  render() {
    const { children } = this.props
    const { index } = this.state
    const titles = children.map(c => c.props.title)

    return (
      <TabContainer { ...this.props }>
        <TabHeader>
          {
            titles.map((title, i) => (
              <TabItem
                key={ title }
                selected={ i === index }
                onClick={ () => this.handleTabChange(i) }
              >
                { title }
              </TabItem>
            ))
          }
        </TabHeader>
        <TabDivider borderColor={ this.props.borderColor } />
        { children.filter((_, i) => i === index) }
      </TabContainer>
    )
  }
}

Tab.propTypes = {
  borderColor: PropTypes.string,
  children: PropTypes.node.isRequired,
}

Tab.defaultProps = {
  borderColor: 'gray.20',
}

export default Tab
