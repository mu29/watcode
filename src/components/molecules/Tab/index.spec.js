import React from 'react'
import { shallow } from 'enzyme'
import { TabPage } from 'components'
import 'jest-styled-components'
import Tab from '.'

const wrap = (props = {}) => shallow((
  <Tab { ...props }>
    <TabPage title="title1">탭 페이지 1</TabPage>
    <TabPage title="title2">탭 페이지 2</TabPage>
  </Tab>
))

describe('<Tab />', () => {
  it('정상적으로 렌더링된다', () => {
    const wrapper = wrap().dive()
    expect(wrapper).toMatchSnapshot()
  })

  it('자식 컴포넌트들의 제목을 그려준다', () => {
    const wrapper = wrap().dive()
    const firstTab = wrapper.childAt(0).childAt(0).dive()
    const secondTab = wrapper.childAt(0).childAt(1).dive()
    expect(firstTab.contains('title1')).toBe(true)
    expect(secondTab.contains('title2')).toBe(true)
  })

  it('기본적으로 첫 페이지를 표시한다', () => {
    const wrapper = wrap().dive()
    expect(wrapper.contains('탭 페이지 1')).toBe(true)
    expect(wrapper.contains('탭 페이지 2')).toBe(false)
  })

  it('선택되지 않은 탭을 누르면 해당 탭 페이지를 표시한다', () => {
    const wrapper = wrap()
    expect(wrapper).toMatchSnapshot()
    wrapper.find({ selected: false }).simulate('click')
    expect(wrapper.state('index')).toBe(1)
    expect(wrapper).toMatchSnapshot()
  })
})
