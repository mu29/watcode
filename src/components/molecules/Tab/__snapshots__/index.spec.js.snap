// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tab /> 선택되지 않은 탭을 누르면 해당 탭 페이지를 표시한다 1`] = `
<styled.div
  borderColor="gray.20"
>
  <styled.div>
    <Styled(Button)
      key="title1"
      onClick={[Function]}
      selected={true}
    >
      title1
    </Styled(Button)>
    <Styled(Button)
      key="title2"
      onClick={[Function]}
      selected={false}
    >
      title2
    </Styled(Button)>
  </styled.div>
  <styled.div
    borderColor="gray.20"
  />
  <TabPage
    title="title1"
  >
    탭 페이지 1
  </TabPage>
</styled.div>
`;

exports[`<Tab /> 선택되지 않은 탭을 누르면 해당 탭 페이지를 표시한다 2`] = `
<styled.div
  borderColor="gray.20"
>
  <styled.div>
    <Styled(Button)
      key="title1"
      onClick={[Function]}
      selected={false}
    >
      title1
    </Styled(Button)>
    <Styled(Button)
      key="title2"
      onClick={[Function]}
      selected={true}
    >
      title2
    </Styled(Button)>
  </styled.div>
  <styled.div
    borderColor="gray.20"
  />
  <TabPage
    title="title2"
  >
    탭 페이지 2
  </TabPage>
</styled.div>
`;

exports[`<Tab /> 정상적으로 렌더링된다 1`] = `
.c0 {
  width: 100%;
  height: 100%;
}

<div
  className="c0"
>
  <styled.div>
    <Styled(Button)
      key="title1"
      onClick={[Function]}
      selected={true}
    >
      title1
    </Styled(Button)>
    <Styled(Button)
      key="title2"
      onClick={[Function]}
      selected={false}
    >
      title2
    </Styled(Button)>
  </styled.div>
  <styled.div
    borderColor="gray.20"
  />
  <TabPage
    title="title1"
  >
    탭 페이지 1
  </TabPage>
</div>
`;
